include/seqan3/alignment/pairwise/policy/alignment_matrix_policy.hpp

```diff
@@ -0,0 +1,157 @@
+// -----------------------------------------------------------------------------------------------------
+// Copyright (c) 2006-2019, Knut Reinert & Freie Universität Berlin
+// Copyright (c) 2016-2019, Knut Reinert & MPI für molekulare Genetik
+// This file may be used, modified and/or redistributed under the terms of the 3-clause BSD-License
+// shipped with this file and also available at: https://github.com/seqan/seqan3/blob/master/LICENSE.md
+// -----------------------------------------------------------------------------------------------------
+
+/*!\file
+ * \brief Provides seqan3::detail::alignment_matrix_policy.
+ * \author Rene Rahn <rene.rahn AT fu-berlin.de>
+ */
+
+#pragma once
+
+#include <seqan3/alignment/band/static_band.hpp>
+#include <seqan3/alignment/pairwise/detail/alignment_algorithm_cache.hpp>
+#include <seqan3/core/type_traits/basic.hpp>
+#include <seqan3/range/views/slice.hpp>
+
+namespace seqan3::detail
+{
+
+/*!\brief Manages the alignment and score matrix.
+ * \ingroup alignment_policy
+ *
+ * \tparam alignment_algorithm_t The type of the seqan3::detail::alignment_algorithm to be augmented with this policy.
+ * \tparam score_matrix_t The type of the used score matrix.
+ * \tparam trace_matrix_t The type of the used trace matrix.
+ *
+ * \details
+ *
+ * This policy is used to manage the score and trace matrix of the alignment algorithm. On invocation of an alignment
+ * instance the necessary memory is allocated and the corresponding matrix iterators are initialised. These
+ * iterators are used as a global state within this particular alignment instance and are accessed from the alignment
+ * algorithm.
+ */
+template <typename alignment_algorithm_t, typename score_matrix_t, typename trace_matrix_t>
```

Comment:
In the documentation you say that alignment_algorithm_t is a specialisation of `seqan3::alignment_algorithm`, why is this not reflected in the code? I thought we avoid concept checks in private member function but not in just any detail struct/function?
(and the template arguments of a class should be checked via static_asserts instead of concepts here then)

Same goes for score_type and trace_type, are they arithmetic?

Reply:
There is no concept for score_matrix_t or trace_matrix_t, yet.

Reply:
algorithm_t cannot be constraint. This is the derived class of this CRTP policy and thus is not fully defined at this point. I will make it clear in the documentation that this is the derived type.